zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -202,21 +210,23 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
                 )
                 .to_word(),
             );
-        });
```

Comment:
I think this is needed.  Notice that here we're checking against `call_callee_address`:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/c40d3da4bf1f4b14ac3b9f607949eb342ed0dfe3/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L202-L212

And `caller_nonce_hash_bytes` is constrained via:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/c40d3da4bf1f4b14ac3b9f607949eb342ed0dfe3/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L213-L217
which calls:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3bbc757a20d9d9f7759db47c096d5395361bee74/zkevm-circuits/src/evm_circuit/util/math_gadget/rlp.rs#L327

Now notice how `input_rlc` uses:
- `self.caller_address` (via `self.caller_address_rlc()`)
- `self.nonce`
Which are unconstrained at the construction of the gadget:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3bbc757a20d9d9f7759db47c096d5395361bee74/zkevm-circuits/src/evm_circuit/util/math_gadget/rlp.rs#L219-L220

So we must check that they are correct.

This checks the nonce (your deleted code):
```rust
        cb.require_equal(
            "tx nonce equivalence",
            tx.nonce.expr(),
            create.caller_nonce(),
        );
```
And this checks the address:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/c40d3da4bf1f4b14ac3b9f607949eb342ed0dfe3/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L188-L192

Reply:
Sorry I just saw that the removed nonce check was moved to https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/c40d3da4bf1f4b14ac3b9f607949eb342ed0dfe3/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L194-L198

So I'm a bit confused.

Are you talking about the removal of the nonce check?  Or something else?

I think it would be cleaner if all the `create` fields equality checks are inside the condition:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/c40d3da4bf1f4b14ac3b9f607949eb342ed0dfe3/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L200-L201

Reply:
There were two `cb.condition(tx.is_create.expr())`:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3bbc757a20d9d9f7759db47c096d5395361bee74/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L193
and the other one:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3bbc757a20d9d9f7759db47c096d5395361bee74/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L213
> I think it would be cleaner if all the create fields equality checks are inside the condition:
I think that's what I wanted to do as well, that's why I merged them.

---

But the second thing I asked is do we still need the constraint from here: https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1475#issue-1758578548
> The callee address of the new context should be the keccak(RLP([sender_address, sender_nonce]))

I think we already have that, right?
cc. @ed255

Reply:
> There were two `cb.condition(tx.is_create.expr())`:
> 
> https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3bbc757a20d9d9f7759db47c096d5395361bee74/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L193
> 
> and the other one:
> 
> https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3bbc757a20d9d9f7759db47c096d5395361bee74/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L213
> 
> > I think it would be cleaner if all the create fields equality checks are inside the condition:
> > I think that's what I wanted to do as well, that's why I merged them.

Ah, thanks for the clarification!  On the constraint level it's the same, but I think in the code level it looks cleaner.

> But the second thing I asked is do we still need the constraint from here: [#1475 (comment)](https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1475#issue-1758578548)
> 
> > The callee address of the new context should be the keccak(RLP([sender_address, sender_nonce]))
> 
> I think we already have that, right? cc. @ed255

Ah yes, I see now that this point was already done.  I think I just missed this
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3bbc757a20d9d9f7759db47c096d5395361bee74/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L193C1-L205C12
Due to it being apart from all the logic of the `is_create` case.  So yeah, all is good regarding this point :)